Src/IronPython/Runtime/PythonEncoding.cs

```diff
@@ -856,4 +998,105 @@ protected void Throw(byte[] bytesUnknown, int index)
                 => throw new DecoderFallbackException($"'surrogatepass' error handler: not a surrogate character", bytesUnknown, index);
         }
     }
+
+    internal class PythonErrorHandlerEncoding : PythonEncoding {
+        private readonly CodeContext _context;
+        private readonly string _errors;
+
+        public PythonErrorHandlerEncoding(CodeContext context, Encoding encoding, string name, string errors)
+            : base(encoding, new PythonHandlerEncoderFallback(), new PythonHandlerDecoderFallback(), name) {
+            _context = context;
+            _errors = errors;
+        }
+
+        private class PythonHandlerEncoderFallback : PythonEncoderFallback {
+            public override int MaxCharCount => int.MaxValue;
+
+            public override EncoderFallbackBuffer CreateFallbackBuffer() {
+                return new PythonHandlerEncoderFallbackBuffer(this.IsPass1, (PythonErrorHandlerEncoding)this.Encoding);
+            }
+        }
+
+        private class PythonHandlerEncoderFallbackBuffer : PythonEncoderFallbackBuffer {
+            private readonly PythonErrorHandlerEncoding _encoding;
+            private object _handler;
+            private char _lastSeenChar;
+
+            public PythonHandlerEncoderFallbackBuffer(bool isPass1, PythonErrorHandlerEncoding encoding)
+                : base(isPass1, encoding) {
+                _encoding = encoding;
+            }
+
+            public override void PrepareIncrement(char[] data, bool forEncoding) {
+                if (Data != null && Data.Length > 0 && EncodingMode) _lastSeenChar <<= Data[0];
+                base.PrepareIncrement(data, forEncoding);
+            }
+
+            public override Tuple<ReadOnlyMemory<char>, ReadOnlyMemory<byte>> GetFallbackCharsOrBytes(int runeUnknown, int index) {
+                if (_handler == null) {
+                    _handler = LightExceptions.CheckAndThrow(PythonOps.LookupEncodingError(_encoding._context, _encoding._errors));
+                }
+
+                // create the exception object to hand to the user-function...
+                int runeLen = runeUnknown > char.MaxValue ? 2 : 1;
```

Comment:
Maybe overkill, but this comes up in a few places so maybe a helper? I assume it would just get inlined by the JIT.
```C#
GetUtf16SequenceLength(int rune) => rune > char.MaxValue ? 2 : 1;
```
The name is based on the `Rune.Utf16SequenceLength` property.

Reply:
This idea had crossed my mind. But I couldn't decide on a name or where to place it. `GetUtf16SequenceLength` is kind of long but since it is already used...

Reply:
If we had a fake Rune class this would just be `runeUnknown.Utf16SequenceLength`.